content/blog/2025-05-13-abstract-testing-floating-point.md

```diff
@@ -0,0 +1,201 @@
+$$p := \frac{a - b}{c + d}$$
+
+We wish to maximize the final error term in this program. Assuming that $a, b,
+c, d$ are all non-negative, it suffices to:
+
+1. Maximize the error term in the numerator ($a-b$). Recursing further, we
+   realize we wish to
+   - minimize the error term in $b$.
+   - maximize the error term in $a$, and,
+2. Minimize the error term in the denominator ($c + d$). Doing the obvious
+   thing, we realize we need to
+   - minimize the error term in $c$, and,
+   - minimize the error term in $d$.
+
+Once we know which direction to point each $\epsilon$, we can assign the most
+extreme values (either $+u$ or $-u$) to get a pretty good $\epsilon$-trace.
+
+For this program, the $\epsilon$-trace we computed happens to be the
+**worst-case**, which is awesome. This technique generalizes in a pretty
+straightforward manner to a backwards static analysis. [^2] 
+
+## Implementation and Preliminary Evaluation
+To implement this technique, I interpret a program twice:
+
+- under an ideal (infinite-precision) semantics with no floating point error,
+and,
+- under an approximate semantics that adds a trace of $\epsilon$s at run-time.
+
+To evaluate, I selected a few benchmarks from the
+[FPBench](https://fpbench.org/benchmarks.html) I had already written a parser
+and could reuse infrastructure for another project. My weapon of choice was
+OCaml for similar practical considerations. [^3]
+
+
+### Charts and Graphs
+Below is a log-scale violin plot showing the distribution of absolute error
+abstractly witnessed by my prototype tool. [^4] (_Aside: I think more people should use
+violin plots._) 
+
+![Violin plot of "Absolute error (abstractly) witnessed by FPBench
+benchmark".](/blog/sample-violins.svg)
+
+Finally, here is a table showing the minimum and maximum error sampled on a few
+selected benchmarks along with the corresponding guarantees provided by FPTaylor
+and Daisy:
+
+| Name | Min. error sampled | Max. error sampled | FPTaylor Guarantee | Daisy Guarantee |
+|:----:|:-----------------:|:-----------------:|:-:|:-:|
+| rigidBody1 | 0.0 | 3.409494-13 | 2.948752e-13 | 2.948752e-13 |
+| rigidBody2 | 0.0 | 3.269484-11 | 3.574474e-11 | 3.606626e-11 |
+| kepler0 | 2.572675e-14 | 6.876940e-14 | 3.463896e-14 | 1.044053e-13|
+| kepler1 | 7.257717e-14 | 2.357817e-13 | 3.689493e-13 | 4.806242e-13 |
+| kepler2 | 5.341705e-13 | 1.628888e-12 | 2.199272e-12 | 2.464839e-12|
+| delta | 5.457664e-13 | 1.657179e-12 |2.197089e-12| 2.346965e-12 |
+| delta4 | 3.126935e-14 |  7.394990e-14 |7.676607e-14| 1.160113e-13|
+
+As you can see, the maximum error sampled can get quite close to the guarantee
+provided by the various tools.
+
+For future work, it would be cool to run this on more tools and benchmarks.
+Additionally, a question not addressed by this tool (and omitted blog post) is
+how to find a good input $x$. Currently, the testing tool uniformly samples from
+the input space. One technique I wrote in my initial proposal (but did not have
```

Comment:
Ah, maybe this answers my question above. But it seems worth hoisting earlier…

Reply:
Done!